server/src/main/java/org/elasticsearch/repositories/RepositoryData.java

```diff
@@ -205,7 +231,10 @@ public RepositoryData removeSnapshot(final SnapshotId snapshotId) {
             indexSnapshots.put(indexId, set);
         }
 
-        return new RepositoryData(genId, newSnapshotIds, newSnapshotStates, indexSnapshots);
+        return new RepositoryData(genId, newSnapshotIds, newSnapshotStates, indexSnapshots,
+            ShardGenerations.builder().addAll(shardGenerations).addAll(updatedShardGenerations)
```

Comment:
Maybe we could build thing alongside with `indexSnapshots` in the `for` loop and avoid the extra `retainIndices` method on the builder that is only used here?

Reply:
Then we have to somehow expose iteration over the generations object. I tried hiding the logic for that a little here. I know it's slightly more code but it seems a little easier to follow?